internal/resources/metal/virtual_circuit/resource.go

```diff
@@ -0,0 +1,459 @@
+		errs = append(errs, d.Set("status", vc.VlanVirtualCircuit.GetStatus()))
+		errs = append(errs, d.Set("nni_vlan", vc.VlanVirtualCircuit.GetNniVlan()))
+		errs = append(errs, d.Set("vnid", vc.VlanVirtualCircuit.GetVnid()))
+		// TODO: this attribute isn't mentioned in the spec
+		errs = append(errs, d.Set("nni_vnid", vc.VlanVirtualCircuit.AdditionalProperties["nni_vnid"]))
+		errs = append(errs, d.Set("name", vc.VlanVirtualCircuit.GetName()))
+		errs = append(errs, d.Set("speed", strconv.Itoa(int(vc.VlanVirtualCircuit.GetSpeed()))))
+		errs = append(errs, d.Set("description", vc.VlanVirtualCircuit.GetDescription()))
+		errs = append(errs, d.Set("tags", vc.VlanVirtualCircuit.GetTags()))
+	} else {
+		errs = append(errs, d.Set("project_id", vc.VrfVirtualCircuit.Project.GetId()))
+		errs = append(errs, d.Set("vrf_id", vc.VrfVirtualCircuit.Vrf.GetId()))
+		errs = append(errs, d.Set("status", vc.VrfVirtualCircuit.GetStatus()))
+		errs = append(errs, d.Set("nni_vlan", vc.VrfVirtualCircuit.GetNniVlan()))
+		// TODO: this attribute isn't mentioned in the spec
+		errs = append(errs, d.Set("nni_vnid", vc.VrfVirtualCircuit.AdditionalProperties["nni_vnid"]))
+		errs = append(errs, d.Set("name", vc.VrfVirtualCircuit.GetName()))
+		errs = append(errs, d.Set("speed", strconv.Itoa(int(vc.VrfVirtualCircuit.GetSpeed()))))
+		errs = append(errs, d.Set("description", vc.VrfVirtualCircuit.GetDescription()))
+		errs = append(errs, d.Set("tags", vc.VrfVirtualCircuit.GetTags()))
+		errs = append(errs, d.Set("peer_asn", vc.VrfVirtualCircuit.GetPeerAsn()))
+		errs = append(errs, d.Set("subnet", vc.VrfVirtualCircuit.GetSubnet()))
+		errs = append(errs, d.Set("metal_ip", vc.VrfVirtualCircuit.GetMetalIp()))
+		errs = append(errs, d.Set("customer_ip", vc.VrfVirtualCircuit.GetCustomerIp()))
+		errs = append(errs, d.Set("md5", vc.VrfVirtualCircuit.GetMd5()))
+	}
+
+	return diag.FromErr(errors.Join(errs...))
+}
+
+func getVCStateWaiter(ctx context.Context, client *metalv1.APIClient, id string, timeout time.Duration, pending, target []string) *retry.StateChangeConf {
+	return &retry.StateChangeConf{
+		Pending: pending,
+		Target:  target,
+		Refresh: func() (interface{}, string, error) {
+			vc, resp, err := client.InterconnectionsApi.GetVirtualCircuit(ctx, id).Execute()
+			if err != nil {
+				if resp != nil {
+					err = equinix_errors.FriendlyErrorForMetalGo(err, resp)
+				}
+				return 0, "", err
+			}
+			vcStatus := ""
+			if vc.VlanVirtualCircuit != nil {
+				vcStatus = string(vc.VlanVirtualCircuit.GetStatus())
+			} else {
+				vcStatus = string(vc.VrfVirtualCircuit.GetStatus())
+			}
+			return vc, vcStatus, nil
+		},
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+}
+
+func resourceMetalVirtualCircuitUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*config.Config).NewMetalClientForSDK(d)
+	needsUpdate := false
+
+	ur := metalv1.VirtualCircuitUpdateInput{}
```

Comment:
I want to think there is some pattern to be used across this resource that would have collapsed the `if vrf {} else vlan {}` blocks in this PR.

That could look like this:

```go
type VCCommon interface {
  GetHref() string
  GetID() string
  Set...
}
```

Where that quickly falls apart is in the custom enum inputs/outputs and the one-off differences between VirtualCircuit types, such as `int` reference vs `*int` for NNID fields.

Withing this Update CRUD, we could apply that pattern. We did so in metal-cli: https://github.com/equinix/metal-cli/blob/main/internal/virtualcircuit/update.go#L12-L22

I only mention this for posterity. Not looking for this PR to adopt this pattern.